modules/videopress/class.videopress-gutenberg.php

```diff
@@ -11,29 +11,69 @@
 class VideoPress_Gutenberg {
 
 	/**
-	 * Initialize the VideoPress Gutenberg extension
+	 * Singleton
 	 */
 	public static function init() {
-		// Should not initialize if Gutenberg is not available or if Jetpack is not active.
-		if (
-			( function_exists( 'register_block_type' ) )
-			&& (
-				( method_exists( 'Jetpack', 'is_active' ) && Jetpack::is_active() )
-				|| ( defined( 'IS_WPCOM' ) && IS_WPCOM )
-			)
-		) {
-			add_action( 'init', array( __CLASS__, 'register_video_block_with_videopress' ) );
+		static $instance = false;
+
+		if ( ! $instance ) {
+			$instance = new VideoPress_Gutenberg();
+		}
+
+		return $instance;
+	}
+
+	/**
+	 * VideoPress_Gutenberg constructor.
+	 *
+	 * Initialize the VideoPress Gutenberg extension
+	 */
+	private function __construct() {
+		add_action( 'init', array( $this, 'set_extension_availability' ) );
+		add_action( 'init', array( $this, 'register_video_block_with_videopress' ) );
+	}
+
+	/**
+	 * Set the Jetpack Gutenberg extension availability.
+	 */
+	public function set_extension_availability() {
+		// It is available on Simple Sites having the appropriate a plan.
+		if ( defined( 'IS_WPCOM' ) && IS_WPCOM ) {
+			if ( '1' === get_option( 'video_upgrade' ) ) {
+				Jetpack_Gutenberg::set_extension_available( 'jetpack/videopress' );
+			} else {
+				Jetpack_Gutenberg::set_extension_unavailable( 'jetpack/videopress', 'missing_plan' );
+			}
+			return;
+		}
+
+		// It is available on Jetpack Sites having the VideoPress module active.
+		if ( method_exists( 'Jetpack', 'is_active' ) && Jetpack::is_active() ) {
+			if ( Jetpack::is_module_active( 'videopress' ) ) {
+				Jetpack_Gutenberg::set_extension_available( 'jetpack/videopress' );
+			} elseif ( ! Jetpack::active_plan_supports( 'videopress' ) ) {
+				Jetpack_Gutenberg::set_extension_unavailable( 'jetpack/videopress', 'missing_plan' );
+			} else {
+				Jetpack_Gutenberg::set_extension_unavailable( 'jetpack/videopress', 'missing_module' );
+			}
```

Comment:
For comparison, this is the check that we use in Simple Payments to determine whether to set the block as available:

https://github.com/Automattic/jetpack/blob/f836eb1f648cb7a8a67573cd208c796c0429a5c0/modules/simple-payments/simple-payments.php#L91-L116

It would be nice to have a helper function for these recurring checks -- ideally by streamlining the WP.com and JP checks for the presence of a given feature. This is probably beyond the scope of this PR -- but for now, do you think it'd make sense to re-use the implementation of that check in Simple Payments (essentially by copy-pasting) to be able to locate these checks more easily later on?

Reply:
When you say re-use the implementation, do you mean to have a `is_enabled_jetpack_videopress` equivalent function that uses `has_any_blog_stickers` for WP.com and `active_plan_supports` for Jetpack?

Actually, my first attempt was to encapsulate the check in a separate `is_enabled_jetpack_videopress` method so we could re-use it later.

But then I realized this is not suitable for VideoPress because `active_plan_supports` returns true if we have the right plan but the module is inactive. Since the `class.videopress-gutenberg.php` file is loaded always, even if the module is not active (https://github.com/Automattic/jetpack/pull/11268), we cannot rely on the default unavailability reason set by `Jetpack_Gutenber::get_availability`:
 https://github.com/Automattic/jetpack/blob/f836eb1f648cb7a8a67573cd208c796c0429a5c0/class.jetpack-gutenberg.php#L357

Also, I remembered the @dmsnell's reminder that good APIs are discovered, not designed (p7H4VZ-29X-p2), so why don't we develop the helper function in the moment we need to re-use the same in a different place?

Reply:
> When you say re-use the implementation, do you mean to have a `is_enabled_jetpack_videopress` equivalent function that uses `has_any_blog_stickers` for WP.com and `active_plan_supports` for Jetpack?

Yeah, something like that for starters -- which we could then extract, parametrize, and re-use in Simple Payments.

> Actually, my first attempt was to encapsulate the check in a separate `is_enabled_jetpack_videopress` method so we could re-use it later.

:heart: 

> But then I realized this is not suitable for VideoPress because `active_plan_supports` returns true if we have the right plan but the module is inactive.

In cases like this, I think it makes sense to ask if these are principle differences, or more of an implementation detail. I like to approach this top-down/semantically: Ideally, we want a helper that can tell us across JP and WP.com if a feature that is part of a paid plan. Modules are similar enough to each other that their activation state shouldn't preclude us from writing such a helper (if anything, the entire concept of a module is something that might not hold up as well in a Gutenberg world, but that's for a different conversation).

So specifically, will it change the block availability behavior of Simple Payments if we add the same module availability check to `is_enabled_jetpack_videopress` that you're using here for VideoPress? (This isn't something you'd need to do as part of this PR -- it's more about starting to converge existing, specific, helper functions, into a shared, generic one.)

> Since the `class.videopress-gutenberg.php` file is loaded always, even if the module is not active (#11268), we cannot rely on the default unavailability reason set by `Jetpack_Gutenber::get_availability`:
> [jetpack/class.jetpack-gutenberg.php](https://github.com/Automattic/jetpack/blob/f836eb1f648cb7a8a67573cd208c796c0429a5c0/class.jetpack-gutenberg.php#L357)
> 
> Line 357 in [f836eb1](/Automattic/jetpack/commit/f836eb1f648cb7a8a67573cd208c796c0429a5c0)
> 
>  				$reason = isset( self::$availability[ $extension ] ) ? self::$availability[ $extension ] : 'missing_module'; 
> Also, I remembered the @dmsnell's reminder that good APIs are discovered, not designed ([p7H4VZ-29X-p2](https://href.li/?https://wp.me/p7H4VZ-29X)), so why don't we develop the helper function in the moment we need to re-use the same in a different place?

I agree with Dennis here, but I'm not sure I'm reading that the way you meant it -- IMO that moment has come, now that both Simple Payments and VideoPress need the same kind of feature/plan check :grimacing:

Reply:
Wonderful, thanks @ockham! I have a clearer perspective now :)

> In cases like this, I think it makes sense to ask if these are principle differences, or more of an implementation detail. I like to approach this top-down/semantically: Ideally, we want a helper that can tell us across JP and WP.com if a feature that is part of a paid plan. Modules are similar enough to each other that their activation state shouldn't preclude us from writing such a helper.

That totally makes sense and indeed the module check won't change the behavior of Simple Payments.

Let me give it a thought and see if I can change this function so we can start reusing it on future iterations.

> IMO that moment has come, now that both Simple Payments and VideoPress need the same kind of feature/plan check

I see it now :) I was considering them different checks!

Reply:
In b7659fc, I encapsulated the check logic in a separated `check_videopress_availability` method. 

I have not created this as a global helper because it needs to be also available in WordPress.com but I'm not able to find any file synced with Fusion that is a good candidate to place this function (all the synced files are hard related to specific modules).

We should create a new file where we can include these helpers functions and add it to Fusion, but since this PR has already a counterpart a diff created by Fusion, I preferred to leave that for a follow-up PR.

Reply:
Actually, I did some tests and I'm afraid that there are some differences between VideoPress and Simple Payments that won't allow us to re-use the helper with the current implementation.

In Jetpack, when I add a Premium Subscription, Jetpack::is_module_active( 'simple-payments' ) returns `false`. But the module is actually enabled and I can add Simple Payments buttons. This means that right now we cannot rely on `is_module_active` for Simple Payments but we need it for VideoPress.

@rodrigoi do you know why the Simple Payments module doesn't get activated in Jetpack?

Reply:
> do you know why the Simple Payments module doesn't get activated in Jetpack?

`simple-payments` is not a module I am afraid. It's one of the "theme tools" that are loaded regardless of module status, here:
https://github.com/Automattic/jetpack/blob/661d73f2f4da09d8520274fa3b641b97fc30db90/modules/module-extras.php#L25

Reply:
Thanks @jeherve.

In that case we need to perform different checks and then I'm not sure if we should go ahead with the re-usable helper. 

We could work-around these differences by indicating via params if we should check the module or not:

```php
function jetpack_check_module_availability( $module, $check_module_is_active = true ) {
	if ( method_exists( 'Jetpack', 'is_active' ) && Jetpack::is_active() ) {
		if ( $check_module_is_active && Jetpack::is_module_active( $module ) ) {
			return array( 'available' => true );
		} elseif ( ! $check_module_is_active && Jetpack::active_plan_supports( $module )  ) {
			return array( 'available' => true );
		} elseif ( ! Jetpack::active_plan_supports( $module ) ) {
			return array(
				'available'          => false,
				'unavailable_reason' => 'missing_plan',
			);
		} else {
			return array(
				'available'          => false,
				'unavailable_reason' => 'missing_module',
			);
		}
	}

	//...
}
```

But these are the kind of abstractions that can lead to a code more difficult to debug.

Reply:
At this point I wonder if we should reconsider our approach, and make Simple Payments a module that we auto-activate when you purchase a plan. It may make things easier to handle.